rio_tiler/io/base.py

```diff
@@ -99,3 +96,255 @@ def preview(self, **kwargs: Any) -> Tuple[numpy.ndarray, numpy.ndarray]:
     def point(self, lon: float, lat: float, **kwargs: Any) -> List:
         """Read a value from a Dataset."""
         ...
+
+
+@attr.s
+class MultiBaseReader(BaseReader, metaclass=abc.ABCMeta):
+    """MultiBaseReader Reader."""
+
+    reader: Type[BaseReader] = attr.ib()
+    reader_options: Dict = attr.ib(factory=dict)
+    bounds: Tuple[float, float, float, float] = attr.ib(init=False)
+    assets: Sequence[str] = attr.ib(init=False)
+
+    @abc.abstractmethod
+    def __enter__(self):
+        """Support using with Context Managers."""
+        ...
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Support using with Context Managers."""
+        pass
+
+    @abc.abstractmethod
+    def _get_asset_url(self, asset: str) -> str:
+        """Validate asset name and construct url."""
+        ...
+
+    def parse_expression(self, expression: str) -> Sequence[str]:
+        """Parse rio-tiler band math expression."""
+        _re = re.compile("|".join(sorted(self.assets, reverse=True)))
```

Comment:
because of regex 
```
import re
_re = re.compile("B1|B11") 
set(re.findall(_re, "B11*B1")) 

> {'B1'}

_re = re.compile("B11|B1") 
set(re.findall(_re, "B11*B1"))
> {'B1', 'B11'}

Reply:
Lol that's a super weak regex... so reverse sorting just happens to put the longer strings first...

It would be much better with fewer edge cases to just improve the regex directly, here the `\b` requires that `B1` and `B11` have a word boundary around them. So `B1` won't match `B11` nor `AB1` nor `B1A` etc
```py
import re
_re = re.compile(r"\bB1\b|\bB11\b")
re.findall(_re, "B11*B1")
# ['B11', 'B1']
```

Reply:
> Lol that's a super weak regex... 

@kylebarron if you want to make fun of code I wrote you can also check https://github.com/vincentsarago/lambda-proxy ;-) 

I agree it's weak and I'm happy make it better ;-)